WIP: Census

abettermap opened the pull request:
<!-- PLEASE REMOVE ANY INAPPLICABLE SECTIONS! -->

## Issues resolved by this pull request

- Closes #113 
- Closes #114
- Closes #115 
- Closes #135

abettermap commented:
@rperlin-ela @fiddleHeads 

Total WIP but in just a few hours I was able to add the PUMA from Maya's MB upload, populate the dropdown directly from the field names (as opposed to config), and reuse the symbology/classes from tract languages: 

https://deploy-preview-134--languagemapping.netlify.app/spatial

i'm sure there are 1000 things not working or accurate, but just an example of how easy it is to work with custom data vs. MB Boundaries, and to populate a dropdown based dynamically off the custom data's fields.

more tomorrow, let me know if i'm on the right track (either here or email, it's pretty redundant topics with this PR).

rperlin-ela commented:
This is awesome— getting close to the dream!

> On Nov 19, 2020, at 1:41 AM, Jason Lampel <notifications@github.com> wrote:
> 
> 
> @rperlin-ela <https://github.com/rperlin-ela> @fiddleHeads <https://github.com/fiddleHeads>
> Total WIP but in just a few hours I was able to add the PUMA from Maya's MB upload, populate the dropdown directly from the field names (as opposed to config), and reuse the symbology/classes from tract languages:
> 
> https://deploy-preview-134--languagemapping.netlify.app/spatial <https://deploy-preview-134--languagemapping.netlify.app/spatial>
> i'm sure there are 1000 things not working or accurate, but just an example of how easy it is to work with custom data vs. MB Boundaries, and to populate a dropdown based dynamically off the custom data's fields.
> 
> more tomorrow, let me know if i'm on the right track (either here or email, it's pretty redundant topics with this PR).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Language-Mapping/language-map/pull/134#issuecomment-730164411>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMNKB5DYGP7JNFK3HG435BTSQS4YPANCNFSM4TX7JSEQ>.
>

abettermap commented:
@rperlin-ela @fiddleHeads `@matt? haha`

good progress tonight, got the PUMA and tracts into one dropdown and all coming from MB tilesets, totally free of Boundaries for census stuff now: https://deploy-preview-134--languagemapping.netlify.app/spatial

plenty of kinks to work out but since i've confirmed that the column order and headings are preserved verbatim, i think as soon as you two work out the fields we can merge this into prod after i get these issues sorted out:

### 1. Spatial panel
the wording and UI in general need some thought:

![image](https://user-images.githubusercontent.com/4974087/99869417-21599e00-2b88-11eb-81bb-a113967f3493.png)

i'll definitely need some input on this, especially for disclaimers and deciding how much of that go into this panel vs. About/Help, etc. Ross and i solidified some things in #135 but there will be more than that.

### 2. Supporting all extents
One of the reasons MB vector tiles work so efficiently is that they _only render and respond to what is needed in the current view_. This is why you see the "zoomed way out initially, then fly in" when the page first loads- it's how I'm able to store the full set of language points. I've researched this and consistently came up with the same conclusion, so I'll need to do something similar with the census features in case a user turns on the census layer when they are zoomed in too close, otherwise they'll end up with the Oil Spill:

![image](https://user-images.githubusercontent.com/4974087/99869511-cc6a5780-2b88-11eb-8b5a-14564be0f000.png)

![image](https://user-images.githubusercontent.com/4974087/99869529-ead05300-2b88-11eb-94e4-7bff5ed67cee.png)

...whereas if they turn it on from full extent, no problemo:

![image](https://user-images.githubusercontent.com/4974087/99869551-308d1b80-2b89-11eb-8af7-85213a1c1f58.png)

### 3. Popups
Not sure it has to be _this branch_, but at some point I'll get the popups set up. I'm thinking another branch would be just fine as i've gotten the majority of functionality covered in the current branch.

### 4. Other?
Aside from those three things, and your data changes, **is there anything else you can see that needs to happen on my end before it can go into production?**

rperlin-ela commented:
I think once the wording and the data are in, we’ll pretty much be good, leaving pop-ups to another round.

Not sure I understand the “Heads up” — right now I’m toggling on Neighborhoods and seeing tracts/PUMAs. I could see it being misleading, but it’s also kind of neat and useful.

In terms of wording, there may be something better but “Spatial” is not bad for the bottom, but do we need a header at the top where it currently says “Spatial dumping ground”? Instinctively I would have said yes, but we don’t have anything for Details or Home and maybe it’s simpler just to have our three self-explanatory things one after another.

Whether or not we try to find some catch-all title, my instinct is to put “Search by location” at the top, “Zoom to my location” after that, then Census at the bottom. I don’t think Census stuff will get lost since we’re sprinkling it elsewhere in the UI. If it’s not too wordy, here’s a suggested bit for under “Census data (NYC only)”— if too wordy, we can use some of it elsewhere:

The Census Bureau’s American Community Survey (ACS), while recording far fewer languages than ELA, provides a useful indication of where the largest several dozen languages are distributed. 

Find below 5-year ACS estimates on “language spoken at home for the Population 5 Years and Over” at the tract level when available, otherwise at the less granular PUMA level. For best results, use together with ELA data [DATA]. ELA is not responsible for Census data or categories. More info here [ABOUT].



> On Nov 21, 2020, at 1:42 AM, Jason Lampel <notifications@github.com> wrote:
> 
> 
> @rperlin-ela <https://github.com/rperlin-ela> @fiddleHeads <https://github.com/fiddleHeads> @matt? haha
> 
> good progress tonight, got the PUMA and tracts into one dropdown and all coming from MB tilesets, totally free of Boundaries for census stuff now: https://deploy-preview-134--languagemapping.netlify.app/spatial <https://deploy-preview-134--languagemapping.netlify.app/spatial>
> plenty of kinks to work out but since i've confirmed that the column order and headings are preserved verbatim, i think as soon as you two work out the fields we can merge this into prod after i get these issues sorted out:
> 
> 1. Spatial panel
> 
> the wording and UI in general need some thought:
> 
>  <https://user-images.githubusercontent.com/4974087/99869417-21599e00-2b88-11eb-81bb-a113967f3493.png>
> i'll definitely need some input on this, especially for disclaimers and deciding how much of that go into this panel vs. About/Help, etc. Ross and i solidified some things in #135 <https://github.com/Language-Mapping/language-map/issues/135> but there will be more than that.
> 
> 2. Supporting all extents
> 
> One of the reasons MB vector tiles work so efficiently is that they only render and respond to what is needed in the current view. This is why you see the "zoomed way out initially, then fly in" when the page first loads- it's how I'm able to store the full set of language points. I've researched this and consistently came up with the same conclusion, so I'll need to do something similar with the census features in case a user turns on the census layer when they are zoomed in too close, otherwise they'll end up with the Oil Spill:
> 
>  <https://user-images.githubusercontent.com/4974087/99869511-cc6a5780-2b88-11eb-8b5a-14564be0f000.png>
>  <https://user-images.githubusercontent.com/4974087/99869529-ead05300-2b88-11eb-94e4-7bff5ed67cee.png>
> ...whereas if they turn it on from full extent, no problemo:
> 
>  <https://user-images.githubusercontent.com/4974087/99869551-308d1b80-2b89-11eb-8af7-85213a1c1f58.png>
> 3. Popups
> 
> Not sure it has to be this branch, but at some point I'll get the popups set up. I'm thinking another branch would be just fine as i've gotten the majority of functionality covered in the current branch.
> 
> 4. Other?
> 
> Aside from those three things, and your data changes, is there anything else you can see that needs to happen on my end before it can go into production?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Language-Mapping/language-map/pull/134#issuecomment-731518405>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMNKB5GPQR76E5IMMDC5443SQ5OLTANCNFSM4TX7JSEQ>.
>

abettermap commented:
> Not sure I understand the “Heads up” — right now I’m toggling on Neighborhoods and seeing tracts/PUMAs. I could see it being misleading, but it’s also kind of neat and useful.

emphasis on "will be" mutually exclusive- haven't coded it yet 😉 

> I could see it being misleading, but it’s also kind of neat and useful.

what, the "heads up" or the mutually exclusive functionality? i will adjust or remove the heads up as needed of course, but are you thinking leave the Neighborhoods/Counties checkbox as-is and just don't worry about coding the mutual exclusivity? fine by me for several reasons:

1. user has ability to hide census layer whenever they want, so they have an easy escape hatch if showing Neighborhoods/Counties and census at the same time becomes too messy. 
1. census stuff is more important so it comes first in the panel, so user might not even get to Neighborhoods/Counties anyway. either way, i'd like to bump the "zoom to my location" section down to the bottom as it's not useful for non-NYCers. that's how it is in the prod map, i must have swapped it accidentally in the shuffle. **NOTE: i hadn't yet read your whole comment yet**
1. i feel like an on/off option for all layers is semi-standard for web maps and it mirrors the logic and behavior of desktop GIS that i find intuitive.
1. it could reduce confusion and the need to explain the mutual exclusivity.
1. census and non-census layers are largely unrelated, so mutual exclusivity seems acceptable.
1. leaving it as-is would save me hours of coding.

> In terms of wording, there may be something better but “Spatial” is not bad for the bottom

if you think of something better, let me know. it IS spatial, but it's more like "reference, bonus, supplemental, extra" to me. if those terms stir up any ideas for you (or Maya), holler.

> but do we need a header at the top where it currently says “Spatial dumping ground”? 

ha, nope! i couldn't think of a good heading so no need to force it. will remove.

> Instinctively I would have said yes, but we don’t have anything for Details or Home and maybe it’s simpler just to have our three self-explanatory things one after another.

good points, agreed overall. if we had a really good heading and intro, sure, but that just seems like fluff and filler if there are already smaller headings and intros further down the panel.

> Whether or not we try to find some catch-all title, my instinct is to put “Search by location” at the top, “Zoom to my location” after that, then Census at the bottom. 

interesting, i would have thought for sure you wanted census first given all the emphasis and effort we're giving it. just curious why you'd want want the location search first? it's mostly just a get-your-bearings convenience while census is a full-blown pattern discovery means.

> I don’t think Census stuff will get lost since we’re sprinkling it elsewhere in the UI.

perhaps, but the sprinkles will be subtle and blended in with surrounding info and actionable buttons, links, etc. Putting it "above the fold", front and center in the panel, would increase the likelihood of use. And if it’s just a matter of making it fit, that shouldn't be a problem once i remove the panel heading.

> If it’s not too wordy, here’s a suggested bit for under “Census data (NYC only)”— if too wordy, we can use some of it elsewhere:

Sounds good, i'll slim it down a little and provide my take then get your review. should be plenty of room either way. questions:

1. regardless of your response to the mutual exclusivity discussion, the "heads up" does not belong below the dropdown list, so that opens up some real estate. what could go there? maybe a one-line PUMA vs. tracts distinction?
1. we've got several UI options in our arsenal: paragraphs, "i" btn popout toggles, below-dropdown label, above-dropdown label, and the full about/help view. could some of your description be broken out into those items or does it make the most sense in 1-2 paragraph format?
1. if we really want this panel to be concise, should we consider an "i" btn toggle for the location and my-location blurbs as well (at least for the my-location), or is that TOO concise? many non-academic non-power users cannot be bothered by reading things, especially reasonably intuitive common stuff like place search and zoom-to-location.
1. if "i" btn for location search blurb, could also change the "Enter a location" placeholder to "Enter a metro-NYC location, e.g. Staten Island" and make it full width?
1. if we did take the ultra-concise road, i think we could scrap the section headings for the location/my-location sections, or maybe lump them together under a "Location Tools" or something?
1. can we change "Show by" above the dropdown to "Show census data by:" or "Choose a census language:" perhaps?

We've got a pretty clean setup going for the other panels, very minimal wording without compromising intuitive use, so I think we could follow suit with Spatial a bit better than we are. I didn't really hit that too hard in the pre-Spatial days because it was a popout in the map, very ephemeral, and sans census. But now with Spatial, more UI stuff to consider and all on a non-ephemeral basis. Btw I never really asked for feedback after the implementation on the whole popover-to-panel migration for Location and My Location, hope it's well-received! Personally i think the my location thing belongs as its own map btn, but that would require a bunch more UI and code decisions which didn't seem worth it for a non-critical feature like this.

You're usually a fan of clean-and-simple so I'll anticipate you're on board with at least some of my minimalist suggestions, but what are your thoughts?

rperlin-ela commented:
> what, the "heads up" or the mutually exclusive functionality? i will adjust or remove the heads up as needed of course, but are you thinking leave the Neighborhoods/Counties checkbox as-is and just don't worry about coding the mutual exclusivity? 

The functionality. My thought was that there might not be a need to code the mutual exclusivity... and all your reasons only reinforce that! 
 
> interesting, i would have thought for sure you wanted census first given all the emphasis and effort we're giving it. just curious why you'd want want the location search first? 

Ok, I don't have strong feelings, I think location search is important too, but the Census functionality is fixing to be pretty kickass, so I'm good with it at the top. 

> 1. regardless of your response to the mutual exclusivity discussion, the "heads up" does not belong below the dropdown list, so that opens up some real estate. what could go there? maybe a one-line PUMA vs. tracts distinction?

Makes sense to me— not exactly sure how to frame in that tight spot. Just "Census tract level data if available, otherwise PUMA"?

> 2. we've got several UI options in our arsenal: paragraphs, "i" btn popout toggles, below-dropdown label, above-dropdown label, and the full about/help view. could some of your description be broken out into those items or does it make the most sense in 1-2 paragraph format?

I definitely will have more to say about the census data in (I think) both About and Help and maybe even the initial disclaimer— what I was trying to put here is the most critical stuff. I'm not sure, but I'm not necessarily averse to try to sticking the rest in an "i"

Everything from "For best results..." on is a little less critical, and of course the Tract/PUMA bit if that goes into what's now the Heads Up. By all means let's tighten further but I think we still we need this info/disclaimer

**The Census Bureau’s American Community Survey (ACS), while recording far fewer languages than ELA, provides a useful indication of where the largest several dozen languages are distributed. Find below 5-year ACS estimates for “language spoken at home for the Population 5 Years and Over”.**

Could split up so that it's more like a one-sentence disclaimer and then "**Source: 5-year ACS estimates for “language spoken at home for the Population 5 Years and Over**". I was kind of torn as to whether we should say "latest" or "most recent" or even a year, but I guess we can do it About, where it'll be easy to change. 

> 3. if we really want this panel to be concise, should we consider an "i" btn toggle for the location and my-location blurbs as well (at least for the my-location), or is that TOO concise? many non-academic non-power users cannot be bothered by reading things, especially reasonably intuitive common stuff like place search and zoom-to-location.

I think I'm into it

> 4. if "i" btn for location search blurb, could also change the "Enter a location" placeholder to "Enter a metro-NYC location, e.g. Staten Island" and make it full width?

Yes, in for the penny...

> 5. if we did take the ultra-concise road, i think we could scrap the section headings for the location/my-location sections, or maybe lump them together under a "Location Tools" or something?

You mean the headers "Zoom to my location" and "Search by location", right? Yes, I think I'd axe these. Not sure it even needs "Location Tools". It's just like— look, we've got a toggle and we've got a search bar!

> 6. can we change "Show by" above the dropdown to "Show census data by:" or "Choose a census language:" perhaps?

Less sure about this one— "Show census data by" feels redundant if "Census Data" is right overhead", but maybe "Choose a language"?

> Btw I never really asked for feedback after the implementation on the whole popover-to-panel migration for Location and My Location, hope it's well-received! Personally i think the my location thing belongs as its own map btn, but that would require a bunch more UI and code decisions which didn't seem worth it for a non-critical feature like this.

Yeah, I can see it both ways but I think it's working pretty well!

> You're usually a fan of clean-and-simple so I'll anticipate you're on board with at least some of my minimalist suggestions, but what are your thoughts?

abettermap commented:
cool, all makes sense. looking like i'm on the right track here?

![image](https://user-images.githubusercontent.com/4974087/99893291-5541cb80-2c3b-11eb-93cd-ed87c70693f7.png)

the ℹ️  for the bottom two are the blurbs from before, and the top one is:

![image](https://user-images.githubusercontent.com/4974087/99893315-a6ea5600-2c3b-11eb-95c1-f12a73725913.png)

i think the "Learn more" (or whatever we'll call the link to Help/About) will fare better in the main intro paragraph above the census dropdown, and we might be able to drop the "i" for the census stuff entirely. plenty of room to move the puma/tracts definition too. or maybe it's already in the right spot, and just needs better wording. we can work on that later.

structure is in pretty good shape code-wise so should be easy to tweak the little content things as we go. let me know what you think so far. pushed to: https://deploy-preview-134--languagemapping.netlify.app/spatial and should be deployed shortly.

rperlin-ela commented:
It’s coming along, but maybe we went a little too succinct.

You might have been right about “Location Tools” or something similar, let’s try it. At the moment, it looks misleadingly like “Census Data” is the header for all of it, especially because the dropdown (for space reasons, I assume) is closer to the location search tools. And maybe just “Census” instead of “Census Data”?

The “i” next to “Show neighborhoods and counties” seems counter-intuitive, goes with the search bar more than the toggle. I think with that we might have to go back to what we had, unless we want to try super short examples  which NYC-area people might know so that it fits in the search bar, like “Enter a metro-NYC location, e.g. Queens, Yonkers, MoMA, 11238, 10 5th Ave”. (If that fits— maybe address not needed?)

The “i” next to “PUMA level” also feels a little funny, so I agree about dropping the “i” for census stuff, I would bite the bullet and try “For best results…” as a second paragraph. “Tract level…” seems ok as is.

I know it’s kind of a weird label since the whole app/map is localized, but what do you think about “Local” (in the sense of local search and display tools) instead of “Spatial”? 

Anyway, if tweakable later, then I’d say full steam ahead since others might have thoughts on these…


> On Nov 21, 2020, at 11:01 PM, Jason Lampel <notifications@github.com> wrote:
> 
> 
> cool, all makes sense. looking like i'm on the right track here?
> 
>  <https://user-images.githubusercontent.com/4974087/99893291-5541cb80-2c3b-11eb-93cd-ed87c70693f7.png>
> the ℹ️ for the bottom two are the blurbs from before, and the top one is:
> 
>  <https://user-images.githubusercontent.com/4974087/99893315-a6ea5600-2c3b-11eb-95c1-f12a73725913.png>
> i think the "Learn more" (or whatever we'll call the link to Help/About) will fare better in the main intro paragraph above the census dropdown, and we might be able to drop the "i" for the census stuff entirely. plenty of room to move the puma/tracts definition too. or maybe it's already in the right spot, and just needs better wording. we can work on that later.
> 
> structure is in pretty good shape code-wise so should be easy to tweak the little content things as we go. let me know what you think so far. pushed to: https://deploy-preview-134--languagemapping.netlify.app/spatial <https://deploy-preview-134--languagemapping.netlify.app/spatial> and should be deployed shortly.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Language-Mapping/language-map/pull/134#issuecomment-731695061>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMNKB5FHG6L3LSOWIFCK7H3SRCEKDANCNFSM4TX7JSEQ>.
>

abettermap commented:
> It’s coming along, but maybe we went a little too succinct.

ha, i might have to agree there. flew too close to the Simplicity Sun!

> You might have been right about “Location Tools” or something similar, let’s try it. At the moment, it looks misleadingly like “Census Data” is the header for all of it, especially because the dropdown (for space reasons, I assume) is closer to the location search tools.

agreed, will do.

> And maybe just “Census” instead of “Census Data”?

hmm, could go either way on that, but will try sans 'Data'. if there's another way you can think of to phrase it, let me know.

> The “i” next to “Show neighborhoods and counties” seems counter-intuitive, goes with the search bar more than the toggle.

oops, i never caught that, yeah totally belongs with the bar.

> I think with that we might have to go back to what we had

possibly, or maybe the blurb _below_ the bar instead? like "hey we assume you'll hit this panel top-to-bottom and if the search bar isn't intuitive enough then just keep on keepin' on and you'll soon find some blurbal relief". 

alternatives:

1. might try the "i" btn next to the bar?
1. or, instead of examples inside the bar, just say "Enter an address, point of interest, etc." then inside the "i" btn popout (wherever its fate may send it), the rest of the explanation.

i feel like Option 2 is still not as good as a pre-bar blurb tho. i'm probably trying too hard for concision again.

> unless we want to try super short examples  which NYC-area people might know so that it fits in the search bar, like “Enter a metro-NYC location, e.g. Queens, Yonkers, MoMA, 11238, 10 5th Ave”. (If that fits— maybe address not needed?)

perhaps, although i'd say a hard Probably Not on the address part. maybe we could just omit the NYC reference from the blurb, put it into "i" popout along with specific examples, then if user searches for non-NYC location and comes up short, we can rest assured, shaking our heads in "should've clicked that 'i' btn, friend-o" disappointment.

one thing to note about the NYC mention is that simply having familiar NYC examples (not to mention a title/logo with "New York City", and a map focused on that area) should minimize potential for "Kyoto" searches, so if we do settle on an "i" btn for this feature, the NYC reference is probably safe and sound hidden behind a popout.

either way, i've got the tone of what we're trying to convey here so i'll play around with some options.

> The “i” next to “PUMA level” also feels a little funny, so I agree about dropping the “i” for census stuff, I would bite the bullet and try “For best results…” as a second paragraph. 

Ok i'll give that a shot, or maybe the second paragraph could go below the dropdown (and below "Tract level...")?

> I know it’s kind of a weird label since the whole app/map is localized, but what do you think about “Local” (in the sense of local search and display tools) instead of “Spatial”?

not bad, it's better than spatial! in case let's roll with the **"Local" icon** we're using in the table and Details then?

## nerdish to human-speak translation

it just occurred to me how often i refer to HTML elements through their content, or sometimes their name but probably with mixed consistency. so, i'll try to just provide you with legends/keys with definitions periodically so you know what the stuff is called, like so:

![image](https://user-images.githubusercontent.com/4974087/99924728-9cda5d00-2cf8-11eb-9d4a-3da5ce59335a.png)

i wasn't able to easily change `German` on the fly, but that is known as an `option` with a `value` of "German" inside a `select menu`.

usually we're referring to the same thing, but i think you knowing a handful of the real terms would simplify our web vocab a bit, and it only takes me a minute or so to make these screenshots.

abettermap commented:
another question- am i pointing "DATA" here to the full raw dataset sheet or the table?

![image](https://user-images.githubusercontent.com/4974087/99893315-a6ea5600-2c3b-11eb-95c1-f12a73725913.png)

locally i've wired up the "about" link to `/about#census` so once you figure out the WP `id` (maybe [this](https://www.wpbeginner.com/beginners-guide/how-to-easily-add-anchor-links-in-wordpress-step-by-step/#:~:text=Add%20the%20ID%20attribute%20to%20the%20linked%20section,Advanced%20tab%20to%20expand%20it.) will be useful?), make sure the `id` for the Census section in About is `census`. if you'd like a different name (or to have it be in Help instead), let me know. OR, we might already have a spot in `/about#sources`, it's just broken.

i think we can get more specific though, so use `census` as the id unless you see a reason not to.

abettermap commented:
one more, and i think i already know the answer to this one- can we do autocomplete/select instead of regular select?

![image](https://user-images.githubusercontent.com/4974087/99927058-23933800-2d01-11eb-9967-4555adfc87dd.png)

just pretend field name is "Italian" and the group name ("PUMA") is preserved above it during typing, exactly like the `A` group name here:

![image](https://user-images.githubusercontent.com/4974087/99927120-59382100-2d01-11eb-859a-f859e829e725.png)

abettermap commented:
...or if it doesn't take too much Fancypants Coding, could skip the group name as a section heading and instead show it _below_ the language name. basically what we're doing in the Omnibox, but much simpler since we only need to show "PUMA" or "tracts" here, like so but with more style:

**Russian**
_Tracts_

**Jamaican/Creole**
_PUMA_

**Bosnian**
_PUMA_

seems too redundant to me, probably roll with group headings instead? that's easier of the two options code-wise.

rperlin-ela commented:
> another question- am i pointing "DATA" here to the full raw dataset sheet or the table?

the table

> one more, and i think i already know the answer to this one- can we do autocomplete/select instead of regular select?

definitely

> ...or if it doesn't take too much Fancypants Coding, could skip the group name as a section heading and instead show it _below_ the language name. 

I'm open but seems redundant, group headings better? Although the way they float now, good in principle, gets a little confusing and hard to read at least for me

abettermap commented:
float as in the "sticky" group headings? do you think the confusion/illegibility is due to one or more of these?

1. short group headings (some only 1 char)
1. light-colored text
1. changing too frequently? (e.g. because there are so many groups)

if so, none of those will be the case with our setup. 

also this is the same behavior as the omnibox group headings, so if you could point out what's different about that vs. census group potential, i might be able to alleviate the issues. 

or maybe you meant something entirely different by "float"...

rperlin-ela commented:
On both mobile and desktop, when I scroll down it’s like “Tract Level” doesn’t go away



> On Nov 23, 2020, at 12:04 PM, Jason Lampel <notifications@github.com> wrote:
> 
> 
> float as in the "sticky" group headings? do you think the confusion/illegibility is due to one or more of these?
> 
> short group headings (some only 1 char)
> light-colored text
> changing too frequently? (e.g. because there are so many groups)
> if so, none of those will be the case with our setup.
> 
> also this is the same behavior as the omnibox group headings, so if you could point out what's different about that vs. census group potential, i might be able to alleviate the issues.
> 
> or maybe you meant something entirely different by "float"...
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Language-Mapping/language-map/pull/134#issuecomment-732294293>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMNKB5F2M5XA43QQG3W55F3SRKIZDANCNFSM4TX7JSEQ>.
>

abettermap commented:
ohhh gotcha. yeah totally different thing, i should have just sent you the live example instead of screenshot:

https://material-ui.com/components/autocomplete/#grouped

that's what we'll be using if you green-light it.

rperlin-ela commented:
Seems good to me— go for it— assume there will be just the two groups (Tract, PUMA) and that language names (good to see them in!) will be in that column order, not alphabetical


> On Nov 23, 2020, at 12:08 PM, Jason Lampel <notifications@github.com> wrote:
> 
> 
> ohhh gotcha. yeah totally different thing, i should have just sent you the live example instead of screenshot:
> 
> https://material-ui.com/components/autocomplete/#grouped <https://material-ui.com/components/autocomplete/#grouped>
> that's what we'll be using if you green-light it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Language-Mapping/language-map/pull/134#issuecomment-732296977>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMNKB5HNQBO3ECJBFRNY2U3SRKJJNANCNFSM4TX7JSEQ>.
>

abettermap commented:
correct. something happened recently that's causing it to alphabetize again in our dropdowns but there's gotta be a way to fix it as it was nicely not sorting before. i'll figure it out.

Re: 2 groups, i just proposed in the email the possibility of a third "Dubious PUMAs" but let me know if we can simply indicate the dubiousness via `^` or something as a first line of defense.

abettermap commented:
@rperlin-ela 

Made some progress today, here's the status report for the [most recent deploy](https://deploy-preview-134--languagemapping.netlify.app/spatial) but read it first before visiting:

## Heads up
1. The dropdown will fail until the **Languages Config** is publicly visible (we can keep that convo going via email). I've been testing with a copy and using my own API key locally, but as soon as you flip the switch back to public it should work (although you might need to clear your browser cache).
1. The code is still using what Maya uploaded to MB and still relying on the "pretty" column names instead of "original" as I haven't had a chance to set up the new stuff yet. That said, some of the languages may not work yet.
1. _Spanish and English_ definitely won't work at tract-level and should be removed from Languages Config until the tracts table includes them.

## What IS working
1. Sort order, pretty names, autocomplete, all populated via your Languages Config LUT's
1. 

## Data things to document
1. Removed `_` from `_original` since it's not going to be internal
1. Everything is dependent on _column order and existence_. It's very fragile in that sense but I'm not aware of a way to reference columns by their name.
1. To mitigate that, I switched from _Named Ranges_ to _Protected Ranges_. This will trigger a warning if any of the column headings (except `_notes`) on the census LUTs or Config tabs is renamed, moved, removed, etc. Still quite fragile but at least it will give data editors some pause.
1. The tracts LUT has a `complicated` column for schema consistency, so just leave it intact (again, the position thing).
1. I'm not using the Census_Metadata at this point but it was a handy reference so let's keep it intact for now. The new updates might prompt some addl changes as well.

## Remaining TODOs 
Despite some of the columns not being ready, I think we should wrap up the following items in this branch and save the Sheets-instead-of-MB approach for another branch as that's going to be a different animal. 

1. Add a legend-like thing for `* It's complicated`. There's quite a bit going on in that panel so let me know where that should ideally live.
1. After that, I'll clean up the UI a bit. It's pretty haphazard but getting better.

## Subsequent branch/es
1. Anything to do with the migration from MB-only to MB/Sheets hybrid.
1. Change "Spatial" to local, and its icon as well. No problem if another branch, it's low priority.
1. I wired up your `/about#census` (nice work setting that up) but it might have some kinks. No problems on your end, I just had a couple more steps on mine. Also the Back/Close behavior of those dialogs is pretty squirrelly, and I'm not sure how to handle it. It's not quite the same concepts as `/table` stuff, but for now as long as the link is working that's good enough.
1. Anything to do with census stuff that's not in the Spatial panel (popups, Details, Explore, etc.)

## Let me know about:
1. Wording and Spatial UI stuff now that you can see/play around with it.
1. Is anything not working/looking right aside from what I disclaimed above?

Good team effort, coming along well and solid config infrastructure in place!

abettermap commented:
@rperlin-ela try not to hit the deploy until you've made the sheet public, otherwise I might get pummeled with Sentry errors (which I also need to handle more gracefully in the code eventually).

Before I forget, _do any of the PUMA2017 or Tracts2013-2017 have blank values_ or can I assume they are all integers (including 0)? I could see blanks jacking up the stats calculations, but I understand `0 does not equal blank` so just give me a heads-up on that if you have an answer.

abettermap commented:
@rperlin-ela try not to hit the deploy until you've made the sheet public, otherwise I might get pummeled with Sentry errors (which I also need to handle more gracefully in the code eventually).

Before I forget, _do any of the PUMA2017 or Tracts2013-2017 have blank values_ or can I assume they are all integers (including 0)? I could see blanks jacking up the stats calculations, but I understand `0 does not equal blank` so just give me a heads-up on that if you have an answer.

rperlin-ela commented:
I won’t check the deploy until I hear back that things are working with the sheet/folder.

I don’t see any reason why we’d have any true blanks in PUMA2017 (if it’s ok I’m going to rename it PUMA2013-2017 which I believe is more accurate) or Tracts2013-2017. Plenty of zeroes, tho.

I think I just took out English and Spanish hopefully without breaking anything.

It’s on us, but I do think the December data update will be an important test of how fragile or robust the system is. There will be some different languages and language names in a different order, not to mention the data, but I think I see things working out.

In terms of the *It’s Complicated, we can try this text *Census Bureau category, component languages unclear and I think basically we want it right under the dropdown helper text? But I’ll see if anything strikes me when I hit the deploy later.

> On Nov 28, 2020, at 9:37 PM, Jason Lampel <notifications@github.com> wrote:
> 
> 
> @rperlin-ela <https://github.com/rperlin-ela> try not to hit the deploy until you've made the sheet public, otherwise I might get pummeled with Sentry errors (which I also need to handle more gracefully in the code eventually).
> 
> Before I forget, do any of the PUMA2017 or Tracts2013-2017 have blank values or can I assume they are all integers (including 0)? I could see blanks jacking up the stats calculations, but I understand 0 does not equal blank so just give me a heads-up on that if you have an answer.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Language-Mapping/language-map/pull/134#issuecomment-735320119>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMNKB5AJ4NHE2AZQ25DOUA3SSGXWBANCNFSM4TX7JSEQ>.
>

rperlin-ela commented:
Checked the deploy and I think it’s looking great — full steam ahead. On mobile it would be great if it went down, like the omnibox search, instead of up. Autocomplete is great. One more line of space before “Location tools” to give a little sense of separateness? I guess *Census Bureau category, component languages unclear  can be like another line of dropdown helper text maybe after “Tract level…” and a blank line? Still seems like people will just launch into the dropdown, but not sure what we can do.

Instead of “Central Park”, let’s put “Astoria”, and instead of “Statue of Liberty”, let’s put “Yankee Stadium”. With luck they’ll see some more interesting languages… Very tiny thing, but on my phone the font size for those e.g.’s seems a little large.

> On Nov 29, 2020, at 11:23 AM, Ross Perlin <perlin@elalliance.org> wrote:
> 
> I won’t check the deploy until I hear back that things are working with the sheet/folder.
> 
> I don’t see any reason why we’d have any true blanks in PUMA2017 (if it’s ok I’m going to rename it PUMA2013-2017 which I believe is more accurate) or Tracts2013-2017. Plenty of zeroes, tho.
> 
> I think I just took out English and Spanish hopefully without breaking anything.
> 
> It’s on us, but I do think the December data update will be an important test of how fragile or robust the system is. There will be some different languages and language names in a different order, not to mention the data, but I think I see things working out.
> 
> In terms of the *It’s Complicated, we can try this text *Census Bureau category, component languages unclear and I think basically we want it right under the dropdown helper text? But I’ll see if anything strikes me when I hit the deploy later.
> 
>> On Nov 28, 2020, at 9:37 PM, Jason Lampel <notifications@github.com <mailto:notifications@github.com>> wrote:
>> 
>> 
>> @rperlin-ela <https://github.com/rperlin-ela> try not to hit the deploy until you've made the sheet public, otherwise I might get pummeled with Sentry errors (which I also need to handle more gracefully in the code eventually).
>> 
>> Before I forget, do any of the PUMA2017 or Tracts2013-2017 have blank values or can I assume they are all integers (including 0)? I could see blanks jacking up the stats calculations, but I understand 0 does not equal blank so just give me a heads-up on that if you have an answer.
>> 
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub <https://github.com/Language-Mapping/language-map/pull/134#issuecomment-735320119>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMNKB5AJ4NHE2AZQ25DOUA3SSGXWBANCNFSM4TX7JSEQ>.
>> 
>

abettermap commented:
> I won’t check the deploy until I hear back that things are working with the sheet/folder.

all good, confirmed locally and in deploy, check away!

> (if it’s ok I’m going to rename it PUMA2013-2017 which I believe is more accurate)

yep call it whatever you want. i'm referencing sheets via their ID, e.g. this nasty thing in the
URL: `1Ts7CvmlKVqCZs_AJ3hf5x1my2p-vaX7q5HxaXiduiEc`

> I think I just took out English and Spanish hopefully without breaking anything.

yep seems to be fine as i can choose other tract-levels and see the result.

> It’s on us, but I do think the December data update will be an important test of how fragile or
> robust the system is. There will be some different languages and language names in a different
> order, not to mention the data, but I think I see things working out.

i set _Data validation_ for the Config's LUT_PUMA_Fields tab based on the
PUMA2013-2017 column headings (except GEOID of course) using
[this](https://support.google.com/docs/thread/4500192?hl=en&msgid=4500453).

> In terms of the *It’s Complicated, we can try this text *Census Bureau category, component
> languages unclear and I think basically we want it right under the dropdown helper text? But I’ll
> see if anything strikes me when I hit the deploy later.

Sounds good, i'll add that for now and work with placement later.

> On mobile it would be great if it went down, like the omnibox search, instead of up.

yeah it's annoying, i remember this being an issue w/omni and hopefully can just copy the solution.

>  One more line of space before “Location tools” to give a little sense of separateness? I guess
>  *Census Bureau category, component languages unclear  can be like another line of dropdown helper
>  text maybe after “Tract level…” and a blank line? Still seems like people will just launch into
>  the dropdown, but not sure what we can do.

ok lots of text to work with now but i'll see what i can come up with.

> Instead of “Central Park”, let’s put “Astoria”, and instead of “Statue of Liberty”, let’s put
> “Yankee Stadium”. With luck they’ll see some more interesting languages…

good idea. done.

> Very tiny thing, but on my phone the font size for those e.g.’s seems a little large.

yeah it's big, all defaults so far, will reduce.

abettermap commented:
Re: tract/puma definition in group heading- i couldn't send a screenshot in chat so here you go

![image](https://user-images.githubusercontent.com/4974087/100549729-fe875380-3231-11eb-9ad4-45e14816c89e.png)

abettermap commented:
@rperlin-ela I pushed up my changes for the afternoon, I think the layout is coming along for Spatial and the dropdown, let me know what you think.

> On mobile it would be great if it went down, like the omnibox search, instead of up.

this is its own little nightmare. it doesn't work like any of the [examples](https://material-ui.com/components/autocomplete/). might have to punt this one until next branch as there are no settings to override it.

there are likely other mobile quirks as well, like how it should behave when you click an item vs. click away from the menu vs. click the menu arrow, and whether or not text is selected when you open it back up, etc.

the MUI "Autocomplete" component is one of the most complex things in their library so i'll do what i can but no promises.

rperlin-ela commented:
All looking good, nice work. Obviously do what you can.

Two last text tweaks if it’s ok—

For tracts, let’s make it “Tracts are the smallest census unit at which language data is provided, and tract-level data will be used here whenever available"

Under PUMAs, two tiny tweaks to axe the hyphen and add a word: “… are a less granular census unit used here whenever tract-level is unavailable."

> On Nov 29, 2020, at 6:06 PM, Jason Lampel <notifications@github.com> wrote:
> 
> 
> @rperlin-ela <https://github.com/rperlin-ela> I pushed up my changes for the afternoon, I think the layout is coming along for Spatial and the dropdown, let me know what you think.
> 
> On mobile it would be great if it went down, like the omnibox search, instead of up.
> 
> this is its own little nightmare. it doesn't work like any of the examples <https://material-ui.com/components/autocomplete/>. might have to punt this one until next branch as there are no settings to override it.
> 
> there are likely other mobile quirks as well, like how it should behave when you click an item vs. click away from the menu vs. click the menu arrow, and whether or not text is selected when you open it back up, etc.
> 
> the MUI "Autocomplete" component is one of the most complex things in their library so i'll do what i can but no promises.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Language-Mapping/language-map/pull/134#issuecomment-735472249>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMNKB5FGKHVNEBXNDJFRGFLSSLHYBANCNFSM4TX7JSEQ>.
>

abettermap commented:
i pushed up something that seems to work (???). total hack but addresses the root of the problem- on small screens (iphone anyway) when the menu does not have enough room between the top of the screen and the virtual keyboard, it will open upward. so, i did a combination of two hacks:

1. assign a height to the menu
1. force the dropdown to scroll into view when it's opened (it's more or less already "in view" but this pushes it up further).

i imagine there will be some wording change requests so i'm still on the fence for using the metadata sheets config, but if you have any obvious changes that stand out, please add them in [that tab](https://docs.google.com/spreadsheets/d/1Ts7CvmlKVqCZs_AJ3hf5x1my2p-vaX7q5HxaXiduiEc/edit#gid=1263096727). i added two more cols as well.

we've hit the back and forth pretty hard on this section, maybe could grab a second opinion if we're stuck on wording/UI?

abettermap commented:
disregard my comment unless there's more to add, i typed it before i saw yours. no problem on the tweaks, i expected such and wasn't stoked on my wording.

abettermap commented:
bien?

![image](https://user-images.githubusercontent.com/4974087/100556554-66549300-3260-11eb-935b-fdff29459911.png)

rperlin-ela commented:
Cool, I think we’ll be good from here and fingers crossed but I think that most of our choices will pass muster in one big sweep from the team

Dropdown hackery looks good on both Phone and Pad!

> On Nov 29, 2020, at 6:30 PM, Jason Lampel <notifications@github.com> wrote:
> 
> ﻿
> disregard my comment unless there's more to add, i typed it before i saw yours. no problem on the tweaks, i expected such and wasn't stoked on my wording.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

abettermap commented:
if so, once i push that up, i will make the tweaks to use the `original` fields and, at least for this branch, do a GIS join and overwrite the current MB census layers to use those fields as well.

i think we can start fresh on the non-dropdown stuff after that. anything else you can see that's critical to go in this branch/PR besides testing once i do those changes?

glad it's working on ipad, i always neglect that. i'm not sure it's subject to the height thing but either way cool that it works.

rperlin-ela commented:
Perf

> On Nov 29, 2020, at 6:32 PM, Jason Lampel <notifications@github.com> wrote:
> 
> ﻿
> bien?
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

abettermap commented:
https://deploy-preview-134--languagemapping.netlify.app/spatial

@rperlin-ela all wired up with the "original" fields in MB now, and also made some minor wording tweaks to the group headings in the dropdown and added some mobile improvements to it.

aside from a small hiccup that could have been avoided, i think it only took me about 15-20 minutes to:

1. export both of the 2013-2017 layers to CSV
1. join them to their spatial layer in GIS
1. export to geojson
1. upload/replace in MB

so, if that's acceptable for Maya to do that with a frequency of ______ times per _______ , as in however often you need to update the data, including adding new columns or renaming existing ones, then let's just run with that and i don't need to adjust anything to make it entirely Sheets-driven. _Note that I still haven't fixed the "oil spill" problem yet_ but I'm going to see how far I can get with that before resorting to Sheets-driven approach.

also keep in mind that the column order and pretty names are 100% controlled via the Languages Config workbook, and the PUMA/tracts validation comes from the 2013-2017 workbooks, so as long as you're working with those and following the same 4-5 step workflow I did, it should be pretty airtight. 

check it out, it's already doing its job:

![image](https://user-images.githubusercontent.com/4974087/100562147-07027d00-3278-11eb-8846-d4731b10da9b.png)

i deleted the "NAMELSAD" col from Tracts2013-2017, which threw off the validation range, so my bad there, but as soon as i fix the `importrange` and Data Validation, voila:

![image](https://user-images.githubusercontent.com/4974087/100562352-b2133680-3278-11eb-9cda-56d56837a859.png)

we'll want to tighten this all up with some docs of course so hopefully you've got your own rough notes going and we can lean on GH and email convos to fill in the rest.

**When you have a chance to review, let me know if it's sufficient to merge into prod** and we can start fresh after that.

rperlin-ela commented:
I’m loving it (including the “x” to clear the dropdown selection I hadn’t noticed before) — merge into prod!

I think we’re still talking at most about one time per year, and I'll ask Maya or Matt to go through this when we have the new December data

Re: the oil spill problem, is that what’s causing the pop-up for Judeo-Kashani (just to take this example) to be so far from its dot? I see this sometimes, not but always, maybe mostly on desktop

 

> On Nov 29, 2020, at 9:31 PM, Jason Lampel <notifications@github.com> wrote:
> 
> 
> https://deploy-preview-134--languagemapping.netlify.app/spatial <https://deploy-preview-134--languagemapping.netlify.app/spatial>
> @rperlin-ela <https://github.com/rperlin-ela> all wired up with the "original" fields in MB now, and also made some minor wording tweaks to the group headings in the dropdown and added some mobile improvements to it.
> 
> aside from a small hiccup that could have been avoided, i think it only took me about 15-20 minutes to:
> 
> export both of the 2013-2017 layers to CSV
> join them to their spatial layer in GIS
> export to geojson
> upload/replace in MB
> so, if that's acceptable for Maya to do that with a frequency of ______ times per _______ , as in however often you need to update the data, including adding new columns or renaming existing ones, then let's just run with that and i don't need to adjust anything to make it entirely Sheets-driven. Note that I still haven't fixed the "oil spill" problem yet but I'm going to see how far I can get with that before resorting to Sheets-driven approach.
> 
> also keep in mind that the column order and pretty names are 100% controlled via the Languages Config workbook, and the PUMA/tracts validation comes from the 2013-2017 workbooks, so as long as you're working with those and following the same 4-5 step workflow I did, it should be pretty airtight.
> 
> check it out, it's already doing its job:
> 
>  <https://user-images.githubusercontent.com/4974087/100562147-07027d00-3278-11eb-8846-d4731b10da9b.png>
> i deleted the "NAMELSAD" col from Tracts2013-2017, which threw off the validation range, so my bad there, but as soon as i fix the importrange and Data Validation, voila:
> 
>  <https://user-images.githubusercontent.com/4974087/100562352-b2133680-3278-11eb-9cda-56d56837a859.png>
> we'll want to tighten this all up with some docs of course so hopefully you've got your own rough notes going and we can lean on GH and email convos to fill in the rest.
> 
> When you have a chance to review, let me know if it's sufficient to merge into prod and we can start fresh after that.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Language-Mapping/language-map/pull/134#issuecomment-735510333>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMNKB5ADDFZNEUQOYHPCGTLSSL7ZJANCNFSM4TX7JSEQ>.
>

abettermap commented:
not sure what you mean about the popup, can you screenshot it? i can't replicate the issue, on my end it looks like this:

![image](https://user-images.githubusercontent.com/4974087/100565054-dc1c2700-327f-11eb-9e89-9e1a72f9d632.png)

the language points and census data should be unrelated in all ways code-wise so i'm reluctant to blame the oil spill, but if you can repeat the problem then you should definitely create an issue, slap a `type: bug` label on it, and throw in some screenshots.

the point you're referring to is outside the tracts and PUMAs so i won't claim that it's not related. the whole stats thing is really fragile and when i try it in this order:

1. start at Judeo point directly from URL: https://deploy-preview-134--languagemapping.netlify.app/details/440
1. turn on something in spatial

i do get this:

![image](https://user-images.githubusercontent.com/4974087/100565354-9f046480-3280-11eb-9d71-9ee529c45e79.png)

merging though, i'd really like to keep the bugs in a separate branch as they're not happening 24/7. mainly just wanted to to get the data/config infrastructure down, and i think we nailed that. the remaining problems should all be code-related, not data or config.

thanks for review, this was a big one.
